libraries/utils/src/render-lines-on-sp.ts

```diff
@@ -63,58 +70,66 @@ export class ScatterPlotLinesRenderer {
-                    this.ctx.lineWidth = this.lines.widths?.[i] ? i === this.currentLineIdx ? this.lines.widths?.[i] + 2 :
+            let pointFromIdx = this.lines.from[i];
+            let pointToIdx = this.lines.to[i];
+            let changedPoints = false;
+            if (pointFromIdx > pointToIdx) {
+                pointFromIdx = this.lines.to[i];
+                pointToIdx = this.lines.from[i];
+                changedPoints = true;
+            }
+            const pointFrom = this.sp.worldToScreen(this.xAxisCol.get(pointFromIdx), this.yAxisCol.get(pointFromIdx));
+            const aX = pointFrom?.x;
+            const aY = pointFrom?.y;
+            const pointTo = this.sp.worldToScreen(this.xAxisCol.get(pointToIdx), this.yAxisCol.get(pointToIdx));
+            const bX = pointTo?.x;
+            const bY = pointTo?.y;
+            if (aX && aY && bX && bY) {
+                const line = new Path2D();
+                this.paths[i] = line;
+                line.moveTo(aX!, aY!);
+                const color = this.lines.color ?? this.lines.colors?.[i] ? this.lines.colors?.[i] : '0,128,0';
+                const opacity = this.lines.opacity ?? this.lines.opacities?.[i] ? this.lines.opacities?.[i] : 1;
+                this.ctx.strokeStyle = `rgba(${color},${opacity})`;
+                this.ctx.lineWidth = this.lines.width ? this.lines.width : this.lines.widths?.[i] ? i === this.currentLineIdx ? this.lines.widths?.[i] + 2 :
                     this.lines.widths?.[i] : i === this.currentLineIdx ? 3 : 1;
-                    let midPoint: DG.Point = this.midPoint(pointFrom, pointTo);
-                    if (this.numberOfLinesPerPair[pairStr] > 1 || linesPerPairCounter[pairStr] > 1) {
-                        midPoint = this.findPerpendicularPointOnCurve(linesPerPairCounter[pairStr],
-                            pointFrom.x, pointFrom.y, midPoint.x, midPoint.y)
-                    }
-                    line.quadraticCurveTo(midPoint.x, midPoint.y, bX, bY);
-                    if (this.lines.drawArrows?.getBit(i))
-                        this.canvasArrow(line, bX, bY, midPoint.x, midPoint.y);
-                    this.ctx.beginPath();
-                    this.ctx.stroke(line);
-                    this.ctx.closePath();
+                let midPoint: DG.Point = this.midPoint(pointFrom, pointTo);
+                if (this.multipleLinesIndices[i]) {
+                    midPoint = this.findPerpendicularPointOnCurve(this.multipleLinesIndices[i],
+                        pointFrom.x, pointFrom.y, midPoint.x, midPoint.y)
                 }
+                line.quadraticCurveTo(midPoint.x, midPoint.y, bX, bY);
+                if (this.lines.drawArrows ?? this.lines.drawArrowsArr?.getBit(i))
+                    this.canvasArrow(line, changedPoints ? aX : bX, changedPoints ? aY : bY, midPoint.x, midPoint.y);
+                this.ctx.beginPath();
+                this.ctx.stroke(line);
+                this.ctx.closePath();
```

Comment:
And here's the biggest performance improvement: you should batch line draws in one stroke() method. Go ahead and compare the performance (I'm also curious). And I think that to keep the code clean, this means we should probably abandon opacities, lines and widths on the series level. Alternatively, you can keep it, but it will make the code a bit more complex. 

https://web.dev/articles/canvas-performance

Or, actually you can have the best of both worlds if you do stroke() in the loop only if you have line-specific attributes defined.

Reply:
Changed - do stroke inside loop only in only in case of individual line styles.

Reply:
As it was agreed, doing stroke for each separate line because otherwise drawing becomes too slow.